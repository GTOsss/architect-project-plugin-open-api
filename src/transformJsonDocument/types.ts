/**
 * Все возможные значения поля OpenAPIV3.SchemaObject.type кроме array и object.
 * */
type OpenAPIPrimitiveType = 'boolean' | 'number' | 'string' | 'integer';

/**
 * Map-объект для трансформации типов в нужный формат.
 * @example:
 * const typesMap = {
 *     integer: 'number',
 *     string: 'string',
 *     boolean: 'boolean',
 *     anyObject: 'AnyObject', // если в scheme только type: object
 *     '*': 'any', // any если не удалось определить тип
 *   };
 * */
export type TypesMap = Record<OpenAPIPrimitiveType | 'anyObject' | '*', string>;

export type Config = {
  typesMap: TypesMap;
};
