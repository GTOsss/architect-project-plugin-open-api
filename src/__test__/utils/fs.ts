import { parse } from 'yaml';
import { readFileSync } from 'fs-extra';

/**
 * Import yaml file as object
 * @example
 * readYaml(resolve(__dirname, '../../mock/openapi.mock.yaml'))
 * */
export const readYaml = (path: string) => {
  return parse(readFileSync(path, 'utf8'));
};
