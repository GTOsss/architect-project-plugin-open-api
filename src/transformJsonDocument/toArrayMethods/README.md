Рассмотрим пример описания OpenAPI методов и того как их будет обрабатывать `toArrayMethods()`. 

Пример целого Open API документа:
```yaml
paths:
  /products:
    get:
      summary: Получить список продуктов
      description: Возвращает список продуктов, отфильтрованных и отсортированных по заданным параметрам.
      tags:
        - Products
      parameters:
        - name: Authorization
          in: header
          description: Токен авторизации для доступа к API.
          required: true
          schema:
            type: string
        - name: category
          in: query
          description: Категория продуктов для фильтрации.
          required: false
          schema:
            type: string
        - name: sortBy
          in: query
          description: Поле для сортировки результатов.
          required: false
          schema:
            type: string
            enum: [name, price, date]
        - name: sortOrder
          in: query
          description: Порядок сортировки (asc или desc).
          required: false
          schema:
            type: string
            enum: [asc, desc]
      responses:
        '200':
          description: Успешный запрос. Возвращает список продуктов.
          content:
            application/json:
              schema:
                type: array
                items:
                  type: object
                  properties:
                    id:
                      type: integer
                    name:
                      type: string
                    description:
                      type: string
                    price:
                      type: number
                      format: float
                    category:
                      type: string
                    imageUrl:
                      type: string
        '401':
          description: Не авторизован. Требуется токен авторизации.
        '500':
          description: Внутренняя ошибка сервера.
  /product/{id}:
    get:
  # ...
```

Отметим верхне-уровневые поля по которым нужно осуществить проход.   

```yaml
paths: # мап-объект 
  /products: # мап-объект
    get:
      parameters: # массив
        # ...
      responses: # мап-объект
        '200':
          # ...
        '401':
          # ...
        '500':
          # ...
    post:
      parameters: 
        # ...
      responses: 
        '200': 
          # ...
  /product/{id}:
    get:
      parameters:
        # ...
      responses:
        # ...
  # ...
```

Для обхода по всем URL используется такая конструкция кода:
```ts
const methods = Object.entries(jsonDocument.paths).reduce((acc, [path, pathData]) => { /* ... */ })
```
Где:
```yaml
paths:  
  /products: # ← path: '/products'
    # ↓ pathData: { get: {}, post: {}, ... }
    get: 
      parameters: 
        # ...
      responses: 
        '200':
    post:
      parameters:
        # ...
      responses:
        '200': 
    # ↑ pathData ↑
          # ...
```

Для прохода по полям `get/post/post/...` заготовлен заранее определенный массив:
```ts
const httpMethodsArray = [
  OpenAPIV3.HttpMethods.POST,
  OpenAPIV3.HttpMethods.PUT,
  OpenAPIV3.HttpMethods.PATCH,
  OpenAPIV3.HttpMethods.GET,
  OpenAPIV3.HttpMethods.DELETE,
];
```

В коде выглядит так: 
```ts
httpMethodsArray.forEach((method) => {
  const currentMethodData = pathData[method];
  /* ... */ 
})
```

Где `currentMethodData` это часть yaml:
```yaml
paths: 
  /products:
    # для текущего URL могут быть описаны parameters которые будут общими для всех методов  
    post: 
    # ↓ currentMethodData ↓
      parameters: # массив
        # ...
      requestBody: # объект
        # ...
      responses: # мап-объект
        '200':
          # ...
        '401':
          # ...
        '500':
          # ...
    # ↑ currentMethodData ↑
    get:
      # ...
```

### Поле `parameters` и метод `parametersToMap()` 
`parametersToMap()` создает мап-объект такого вида:
```ts
/**
 * // {
 * //   header: [ParameterObject, ...],
 * //   query: [ParameterObject, ...],
 * //   path: [ParameterObject, ...],
 * // }
 * */
const parametersMap = {
  header: [
    {
      in: 'header',
      name: 'Authorization',
      required: true,
      type: 'string', // готовый тип для кодогенерации
    },
  ],
  path: [],
  query: [
    {
      in: 'query',
      name: 'page',
      required: false,
      type: 'number', // готовый тип для кодогенерации
    },
    {
      in: 'query',
      name: 'role',
      type: "'admin' | 'user' | 'guest'", // готовый тип для кодогенерации
    },
  ],
};


```

### Поле `requestBody`

### Поле `responses`

