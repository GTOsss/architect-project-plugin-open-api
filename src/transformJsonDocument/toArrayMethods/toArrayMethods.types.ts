import { OpenAPIV3 } from 'openapi-types';
import HttpMethods = OpenAPIV3.HttpMethods;
import { ParametersMap } from '../utils/parametersToMap/parametersToMap.types';

/**
 * Contain *OpenAPI* and *Arc* fields.
 * - *OpenAPI* fields - native fields from OpenAPI yaml/json document.
 * - *Arc* fields - field prepared for code generator.
 * */
export type Method = {
  /**
   * OpenAPI field.
   * @example GET, POST, PUT, DELETE, etc */
  method: HttpMethods;

  /**
   * OpenAPI field.
   * @example 'users/:id' */
  path: string;

  /**
   * OpenAPI field.
   * Description from openAPI document */
  description: string;

  /**
   * OpenAPI field.
   * Summary from openAPI document */
  summary: string;

  /**
   * OpenAPI field.
   * Tags from openAPI document */
  tags: string[];

  /**
   * Arc field.
   * {@link ParametersMap} */
  parametersMap: ParametersMap;

  /**
   * Arc field.
   * Typescript type generated by scheme of success status or "default"
   * - 200 || 201 || default
   * - "application/json" || "multipart/form-data" || "application/x-www-form-urlencoded"
   * */
  responseType: string;

  /**
   * Arc fields.
   * Typescript type generated by scheme of [path][method].requestBody
   * */
  requestBodyType: string;
};
