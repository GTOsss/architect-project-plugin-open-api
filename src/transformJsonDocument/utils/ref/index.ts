export * from './resolveRef';
export * from './getSchemaNameByRef';
export * from './parseRefPath';
