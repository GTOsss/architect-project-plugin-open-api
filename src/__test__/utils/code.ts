import { existsSync } from 'fs';
import { ESLint } from 'eslint';

type ApplyEslintAutoFixParams = {
  code: string;
  eslintConfigPath: string;
  filePath: string;
};

/**
 * Apply autofix for input code
 * */
export const applyEslintAutofix = async ({
  code,
  eslintConfigPath,
  filePath,
}: ApplyEslintAutoFixParams): Promise<string> => {
  try {
    if (!existsSync(eslintConfigPath)) {
      throw new Error(`ESLint configuration file not found by path: ${eslintConfigPath}`);
    }

    const eslint = new ESLint({
      fix: true,
      overrideConfigFile: eslintConfigPath,
    });

    const results = await eslint.lintText(code, { filePath });

    await ESLint.outputFixes(results);

    return results[0]?.output || code;
  } catch (error) {
    console.error('Error applying eslint autofix:', error);
    return code;
  }
};

export const applyEslint = async (code: string) =>
  await applyEslintAutofix({ code, filePath: 'temp/forEslintAutofix.ts', eslintConfigPath: 'eslint.config.mjs' });
