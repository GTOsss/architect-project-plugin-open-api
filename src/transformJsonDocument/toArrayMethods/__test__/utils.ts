import { OpenAPIV3 } from 'openapi-types';
import { toArrayMethods } from '../toArrayMethods';
import { Method } from '../toArrayMethods.types';
import { TypesMap } from '../../types';

const typesMap: TypesMap = {
  integer: 'number',
  number: 'number',
  string: 'string',
  boolean: 'boolean',
  anyObject: 'AnyObject', // если в scheme только type: object
  '*': 'any', // any если не удалось определить тип
};

/**
 * Находит объект указанного метода.
 * @example findMethod(HttpMethods.GET, 'users/{id}', methods);
 * */
export const findMethod = (method: OpenAPIV3.HttpMethods, path: string, methods: Method[]): Method => {
  return methods.find((currentMethod) => currentMethod.method === method && currentMethod.path === path);
};

/**
 * Вызывает getMethodsAsArray и находит в полученном массиве указанный метод.
 * */
export const testFor = (method: OpenAPIV3.HttpMethods, path: string, jsonDocument: OpenAPIV3.Document): Method => {
  const methods = toArrayMethods(jsonDocument, { typesMap });
  return findMethod(method, path, methods);
};
