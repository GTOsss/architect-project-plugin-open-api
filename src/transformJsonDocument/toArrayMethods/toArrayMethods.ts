import { OpenAPIV3 } from 'openapi-types';
import { isNotDefined } from '#/common/types/common';
import { parametersToMap } from '../utils/parametersToMap';
import { Method } from './toArrayMethods.types';
import { resolveResponseType } from '../utils/resolveResponseType';
import { Config } from '../types';
import { resolveRequestBodyType } from '#/transformJsonDocument/utils/resolveRequestBodyType';
import { methodNameByPath } from '#/transformJsonDocument/utils/pathUtils/methodNameByPath/methodNameByPath';
import { urlGetterByPath } from '#/transformJsonDocument/utils/pathUtils/urlGetterByPath/urlGetterByPath';

/**
 * Для обхода Open API методов.
 * ```yaml
 * paths: # мап-объект
 *   /products: # мап-объект
 *     get: # <<<
 *         # ...
 *     post: # <<<
 *         # ...
 * ```
 * */
const httpMethodsArray = [
  OpenAPIV3.HttpMethods.POST,
  OpenAPIV3.HttpMethods.PUT,
  OpenAPIV3.HttpMethods.PATCH,
  OpenAPIV3.HttpMethods.GET,
  OpenAPIV3.HttpMethods.DELETE,
];

/**
 * 1. Преобразует оригинальный JSON документ open API в массив объектов, где каждый элемент массива это конкретный
 * ендпоинт.
 * 2. Резолвит все $ref объекты
 *
 * Подробное описание логики функции в ./README.md
 *
 * */
export const toArrayMethods = (jsonDocument: OpenAPIV3.Document, config: Config) => {
  // Проход по всем paths
  const methods = Object.entries(jsonDocument.paths).reduce((acc, [path, pathData]) => {
    const pathDataParameters = pathData.parameters || [];

    // Проход по HTTP Methods: "POST users" > "PUT users" > "PATCH users" > "GET users" > ...
    httpMethodsArray.forEach((method) => {
      const currentMethodData = pathData[method];

      if (isNotDefined(currentMethodData)) {
        return;
      }

      const { description, summary, tags, parameters } = currentMethodData;
      const methodDataParameter = parameters || [];

      // Взять параметры текущего метода (GET /users) и общие параметры от текущего path (/users)
      const allParameters = [...methodDataParameter, ...pathDataParameters];

      // Хранит map объект полученный проходом по массиву параметров конкретного REST API.
      const parametersMap = parametersToMap({ parameters: allParameters, jsonDocument, config });

      /**
       * Хранит сгенерированный тип схемы успешного статуса.
       * Поиск ближайшего не пустого значения в OpenAPI документе происходит по следующем полям:
       * - 200 || 201 || ... || default
       * - "application/json" || "multipart/form-data" || "application/x-www-form-urlencoded"
       * */
      const responseType = resolveResponseType(jsonDocument, currentMethodData.responses, config);

      /**
       * Хранит сгенерированный тип схемы поля requestBody
       * */
      const requestBodyType = resolveRequestBodyType(jsonDocument, currentMethodData.requestBody, config);

      const generatedMethodName = methodNameByPath(path);

      const urlGetter = urlGetterByPath(path, method);

      acc.push({
        generatedMethodName,
        path,
        method,
        description,
        summary,
        tags,
        parametersMap,
        responseType,
        requestBodyType,
        urlGetter,
      });
    });

    return acc;
  }, [] as Method[]);

  return methods;
};
