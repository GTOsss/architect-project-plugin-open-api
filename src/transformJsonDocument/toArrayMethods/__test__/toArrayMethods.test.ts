import { testFor } from './utils';
import { OpenAPIV3 } from 'openapi-types';
import { readYaml } from '../../../__test__/utils/fs';
import { resolve } from 'path';
import HttpMethods = OpenAPIV3.HttpMethods;
import { toArrayMethods } from '../toArrayMethods';
import { TypesMap } from '../../types';

const openApiMock = readYaml(resolve(__dirname, '../../../mock/openapi.mock.yaml')) as OpenAPIV3.Document;
const openApiMockResponse200 = readYaml(
  resolve(__dirname, '../../../mock/cases/response200.yaml'),
) as OpenAPIV3.Document;
const openApiMockResponse200Extended = readYaml(
  resolve(__dirname, '../../../mock/cases/response200Extended.yaml'),
) as OpenAPIV3.Document;

const typesMap: TypesMap = {
  number: 'number',
  integer: 'number',
  string: 'string',
  boolean: 'boolean',
  anyObject: 'AnyObject', // если в scheme только type: object
  '*': 'any', // any если не удалось определить тип
};

const config = { typesMap };

describe('toArrayMethods mock/cases', () => {
  it('responseTypes: responses200.yaml', () => {
    const result = toArrayMethods(openApiMockResponse200, config);
    expect(result.map(({ responseType, path }) => ({ path, responseType }))).toMatchSnapshot();
  });

  it('responseTypes: response200Extended.yaml', () => {
    const result = toArrayMethods(openApiMockResponse200Extended, config);
    expect(result.map(({ responseType, path }) => ({ path, responseType }))).toMatchSnapshot();
  });
});

describe('toArrayMethods by openapi.mock.json', () => {
  it('GET /users', () => {
    expect(testFor(HttpMethods.GET, '/users', openApiMock)).toMatchSnapshot();
  });

  it('POST /users', () => {
    expect(testFor(HttpMethods.POST, '/users', openApiMock)).toMatchSnapshot();
  });

  it('PUT /users/{id}', () => {
    expect(testFor(HttpMethods.PUT, '/users/{id}', openApiMock)).toMatchSnapshot();
  });

  it('PATCH /users/{id}', () => {
    expect(testFor(HttpMethods.PATCH, '/users/{id}', openApiMock)).toMatchSnapshot();
  });

  it('DELETE /users/{id}', () => {
    expect(testFor(HttpMethods.DELETE, '/users/{id}', openApiMock)).toMatchSnapshot();
  });
});
