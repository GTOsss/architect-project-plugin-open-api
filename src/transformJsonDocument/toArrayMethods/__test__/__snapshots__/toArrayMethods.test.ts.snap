// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`toArrayMethods by openapi.mock.json DELETE /users/{id} 1`] = `
{
  "description": undefined,
  "method": "delete",
  "parametersMap": {
    "header": [
      {
        "in": "header",
        "name": "Authorization",
        "required": true,
        "type": "string",
      },
    ],
    "path": [
      {
        "in": "path",
        "name": "id",
        "required": true,
        "type": "number",
      },
    ],
    "query": [],
  },
  "path": "/users/{id}",
  "responseType": undefined,
  "summary": "Delete user",
  "tags": undefined,
}
`;

exports[`toArrayMethods by openapi.mock.json GET /users 1`] = `
{
  "description": undefined,
  "method": "get",
  "parametersMap": {
    "header": [
      {
        "in": "header",
        "name": "Authorization",
        "required": true,
        "type": "string",
      },
    ],
    "path": [],
    "query": [
      {
        "in": "query",
        "name": "page",
        "required": false,
        "type": "number",
      },
      {
        "in": "query",
        "name": "role",
        "type": ""admin" | "user" | "guest"",
      },
    ],
  },
  "path": "/users",
  "requestBodyType": undefined,
  "responseType": "Array<User>",
  "summary": "List users",
  "tags": undefined,
}
`;

exports[`toArrayMethods by openapi.mock.json PATCH /users/{id} 1`] = `
{
  "description": undefined,
  "method": "patch",
  "parametersMap": {
    "header": [
      {
        "in": "header",
        "name": "Authorization",
        "required": true,
        "type": "string",
      },
    ],
    "path": [
      {
        "in": "path",
        "name": "id",
        "required": true,
        "type": "number",
      },
    ],
    "query": [],
  },
  "path": "/users/{id}",
  "responseType": "User",
  "summary": "Partially update user",
  "tags": undefined,
}
`;

exports[`toArrayMethods by openapi.mock.json POST /users 1`] = `
{
  "description": undefined,
  "method": "post",
  "parametersMap": {
    "header": [
      {
        "in": "header",
        "name": "Authorization",
        "required": true,
        "type": "string",
      },
    ],
    "path": [],
    "query": [],
  },
  "path": "/users",
  "responseType": "User",
  "summary": "Create user",
  "tags": undefined,
}
`;

exports[`toArrayMethods by openapi.mock.json PUT /users/{id} 1`] = `
{
  "description": undefined,
  "method": "put",
  "parametersMap": {
    "header": [
      {
        "in": "header",
        "name": "Authorization",
        "required": true,
        "type": "string",
      },
    ],
    "path": [
      {
        "in": "path",
        "name": "id",
        "required": true,
        "type": "number",
      },
    ],
    "query": [],
  },
  "path": "/users/{id}",
  "responseType": "User",
  "summary": "Update user",
  "tags": undefined,
}
`;

exports[`toArrayMethods mock/cases responseTypes: responses200.yaml 1`] = `
[
  {
    "path": "/users/me",
    "responseType": "User",
  },
  {
    "path": "/users",
    "responseType": "Array<User>",
  },
]
`;
